src/libostree/ostree-sysroot-deploy.c

```diff
@@ -2456,61 +2452,259 @@ ostree_sysroot_deploy_tree (OstreeSysroot     *self,
+                             GCancellable      *cancellable,
+                             GError           **error)
+{
+  g_autoptr(OstreeDeployment) merge_deployment = NULL;
+  if (provided_merge_deployment != NULL)
+    merge_deployment = g_object_ref (provided_merge_deployment);
+
+  g_autofree char *deployment_path = ostree_sysroot_get_deployment_dirpath (self, deployment);
+  glnx_autofd int deployment_dfd = -1;
+  if (!glnx_opendirat (self->sysroot_fd, deployment_path, TRUE, &deployment_dfd, error))
     return FALSE;
 
+  g_autoptr(OstreeSePolicy) sepolicy = NULL;
+  if (!merge_configuration (self, ostree_sysroot_repo (self),
+                            merge_deployment, deployment,
+                            deployment_dfd, &sepolicy,
+                            cancellable, error))
+    return FALSE;
+
+  /* Don't fsync here, as we assume that's all done in
+   * ostree_sysroot_write_deployments().  We do rewrite the origin
+   * file again after the config merge to be doubly sure that we
+   * have the right SELinux context.
+   */
+  if (!write_origin_file_internal (self, sepolicy, deployment, NULL,
+                                   GLNX_FILE_REPLACE_NODATASYNC,
+                                   cancellable, error))
+    return FALSE;
+
+  if (!selinux_relabel_var_if_needed (self, sepolicy, deployment, cancellable, error))
+    return FALSE;
+
+  /* Seal it */
   if (!(self->debug_flags & OSTREE_SYSROOT_DEBUG_MUTABLE_DEPLOYMENTS))
     {
-      if (!ostree_sysroot_deployment_set_mutable (self, new_deployment, FALSE,
+      if (!ostree_sysroot_deployment_set_mutable (self, deployment, FALSE,
                                                   cancellable, error))
         return FALSE;
     }
 
-  { g_auto(OstreeSepolicyFsCreatecon) con = { 0, };
+  return TRUE;
+}
 
-    if (!_ostree_sepolicy_preparefscreatecon (&con, sepolicy,
-                                              "/etc/ostree/remotes.d/dummy.conf",
-                                              0644, error))
-      return FALSE;
+/**
+ * ostree_sysroot_deploy_tree:
+ * @self: Sysroot
+ * @osname: (allow-none): osname to use for merge deployment
+ * @revision: Checksum to add
+ * @origin: (allow-none): Origin to use for upgrades
+ * @provided_merge_deployment: (allow-none): Use this deployment for merge path
+ * @override_kernel_argv: (allow-none) (array zero-terminated=1) (element-type utf8): Use these as kernel arguments; if %NULL, inherit options from provided_merge_deployment
+ * @out_new_deployment: (out): The new deployment path
+ * @cancellable: Cancellable
+ * @error: Error
+ *
+ * Check out deployment tree with revision @revision, performing a 3
+ * way merge with @provided_merge_deployment for configuration.
+ *
+ * While this API is not deprecated, you most likely want to use the
+ * ostree_sysroot_stage_tree() API.
+ */
+gboolean
+ostree_sysroot_deploy_tree (OstreeSysroot     *self,
```

Comment:
Maybe let's start by splitting out this `initialize` + `finalize` separation in a prep PR?